### PR TITLE
Initial implementation of the library

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: libjcat/jcat-engine-pkcs7.c [hunk 1]
```diff
@@ -0,0 +1,834 @@
...
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "privkey_verify_params: %s [%i]",
+			     gnutls_strerror (rc), rc);
+		return FALSE;
+	}
+
+	/* create parents if required */
+	if (!jcat_mkdir_parent (fn, error))
+		return FALSE;
```
Comment (superm1) [inline libjcat/jcat-engine-pkcs7.c]: I would think make the directory first, then the key, then export it.  That way if the directory can't be made you fail earlier.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,61 @@
-# libjcat
-Library for reading and writing Jcat files
+libjcat
+=======
+
+This library allows reading and writing gzip-compressed JSON catalog files,
+which can be used to store GPG, PKCS-7 and SHA-256 checksums for each file..
+
+This provides equivalent functionality to the catalog files supported in
+Microsoft Windows.
```
[the remark this answers is not in the document]
Comment (hughsie) [reply README.md]: https://blogs.gnome.org/hughsie/2020/02/28/introducing-jcat/ -- MS broke on disk format shortly before Win10 -- and it's not like we can edit the WU-supplied cat file anway.

File: meson.build [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+]
+
+if get_option('pkcs7')
+  gnutls = dependency('gnutls', version : '>= 3.4.4.1')
+  if gnutls.version().version_compare('>= 3.6.0')
+    conf.set('HAVE_GNUTLS_3_6_0', '1')
+  endif
+  conf.set('ENABLE_PKCS7', '1')
+  libjcat_deps += gnutls
+endif
```
[the remark this answers is not in the document]
Comment (hughsie) [reply meson.build]: Very valid! Fixed.

File: libjcat/jcat-engine.c [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+jcat_engine_setup (JcatEngine *self, GError **error)
+{
+	JcatEngineClass *klass = JCAT_ENGINE_GET_CLASS (self);
+	g_return_val_if_fail (JCAT_IS_ENGINE (self), FALSE);
+	if (klass->setup == NULL)
+		return TRUE;
+	return klass->setup (self, error);
+}
+
+gboolean
```
[the remark this answers is not in the document]
Comment (hughsie) [reply libjcat/jcat-engine.c]: private, but added comment to that effect

File: libjcat/jcat-engine.c [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+#define GET_PRIVATE(o) (jcat_engine_get_instance_private (o))
+
+enum {
+	PROP_0,
+	PROP_KIND,
+	VERIFY_KIND,
+	PROP_LAST
+};
+
+gboolean
```
[the remark this answers is not in the document]
Comment (hughsie) [reply libjcat/jcat-engine.c]: private, but added comment to that effect

File: libjcat/jcat-engine-pkcs7.c (hunk 1 above)
Comment (hughsie) [reply libjcat/jcat-engine-pkcs7.c]: Agree, fixed.

File: contrib/ci/Dockerfile-debian [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	libjson-glib-dev \
+	ninja-build \
+	pkg-config \
+	python3-distutils \
+	python3-pip \
+	python3-setuptools \
+	python3-wheel
+
+# Meson is too old in unstable, and that won't change until Buster is released
+RUN pip3 install meson
```
Comment (superm1) [inline contrib/ci/Dockerfile-debian]: Is this true?  What version of meson do you need?

File: contrib/ci/Dockerfile-debian [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+Pin: release a=testing\n\
+Pin-Priority: 900\n\
+\n\
+Package: *\n\
+Pin: release a=unstable\n\
+Pin-Priority: 800\n'\
+ > /etc/apt/preferences
+ > /etc/apt/preferences
+RUN cat /etc/apt/preferences
+RUN echo fubar > /etc/machine-id
```
Comment (superm1) [inline contrib/ci/Dockerfile-debian]: I would think this whole block is probably not needed

File: contrib/ci/Dockerfile-debian (hunk 6 above)
Comment (hughsie) [reply contrib/ci/Dockerfile-debian]: 0.47.0 -- tbh I don't really know what `unstable` and `testing` are...

File: contrib/ci/Dockerfile-debian (hunk 7 above)
Comment (hughsie) [reply contrib/ci/Dockerfile-debian]: Could you do me a favour and fix the Debian CI? It's a bit weird to my little Red Hat brain.

File: contrib/ci/Dockerfile-debian (hunk 6 above)
Comment (superm1) [reply contrib/ci/Dockerfile-debian]: 🤦‍♂ Here's a decoder ring:
Unstable: brand new stuff
Testing: brand new stuff that has passed automatic testing and promption after X days
Stable: Current release

https://tracker.debian.org/pkg/meson

It's new enough in testing and stable.

File: contrib/ci/Dockerfile-debian (hunk 7 above)
Comment (superm1) [reply contrib/ci/Dockerfile-debian]: That was more painful than I expected to get s390x working. But it's there now.

Something is awry in the test suite for Windows still though.

**Pull request merged**